### PR TITLE
start trying to fix the pf_key tests

### PR DESCRIPTION
**Pull request by richlowe**
This doesn't succeed in fixing the tests -- I'm not sure what they need in the way of network env beyond "a network", as I fixed -- but it fixes _something_.

Most trivially, fix PATH for the tests and check we succeeded in finding a network.

Less trivially, fix ipsecconf(8), which in some circumstances will attempt to `strlcpy()` with overlapping buffers and on ARM -- but nowhere else we run, shucks -- fail, as the current ARM memcpy is not overlap safe in that specific path (our "undefined behaviour" smashes the buffer.  i386's is apparently "work").

@danmcd could you look at especially ipsecconf?  That fix should go to illumos-gate, but I don't have a sensible test setup for x86 right now.

### CODE REVIEW

**Comment by danmcd** [comment]
The test needs "A link-with-prefix and a default route."  Exactly one configured {v,}NIC in its netstack, and exactly one default router.

I'd love to see a coredump or at least the input file (or stdin) from one of the ipsecconf strlcpy() failures; would love to know what the input was that made it fail.

**Comment by richlowe** [comment]
The failing input was from https://github.com/illumos/illumos-gate/blob/55fea89dcaa64928bed4327112404dcb3e07b79f/usr/src/test/os-tests/tests/pf_key/acquire-compare.sh#L78

The failures corrupt the buffer so parsing fails semi-arbitrarily.  I'm guessing we know the tests pass so x86 is fine, at least for a 32bit `ipsecconf`.

File: usr/src/cmd/cmd-inet/usr.sbin/ipsecutils/ipsecconf.c [hunk 1]
```diff
@@ -3929,8 +3937,16 @@ parse_pattern_or_prop(FILE *fp, char *argvec[], char **leftover)
 				 * leftover buffer if any.
 				 */
 				if (*buf != '\0') {
-					(void) strlcpy(lo_buf, buf,
+					/*
```
Comment (danmcd) [inline usr/src/cmd/cmd-inet/usr.sbin/ipsecutils/ipsecconf.c]: Suggestion:  Rip this out into a function:

```
/*
 * NB: src may point into part or all of dst. Let's play
 * it safe and use this as a wrapper to memmove().
 * XXX KEBE SAYS could return (len) just like strlcpy().
 */
void
strlcpy_move(char *dst, const char *src, size_t dstlen)
{
        size_t len = MIN(strlen(src), dstlen);
        
        memmove(dst, src, len);
}

File: usr/src/test/os-tests/tests/pf_key/acquire-compare.sh [hunk 2]
```diff
@@ -37,6 +41,12 @@ fi
 
 # NOTE: If multihomed, this may fail in interesting ways...
 MY_IP=`netstat -in -f inet | egrep -v "Name|lo0" | awk '{print $4}' | head -1`
+
```
Comment (danmcd) [inline usr/src/test/os-tests/tests/pf_key/acquire-compare.sh:44]: A further reality check might be to check that TEST_REMOTE_DST* (see below) hit a route, and one that isn't RTF_BLACKHOLE (which returns ENETUNREACH).  Even if they packets go off to hyperspace, the netstack must think they can leave out the {,v}NIC successfully.

File: usr/src/cmd/cmd-inet/usr.sbin/ipsecutils/ipsecconf.c (hunk 1 above)
Comment (richlowe) [reply usr/src/cmd/cmd-inet/usr.sbin/ipsecutils/ipsecconf.c]: Ripped out more thoroughly, in a way that guarantees termination etc.

File: usr/src/test/os-tests/tests/pf_key/acquire-compare.sh (hunk 2 above)
Comment (richlowe) [reply usr/src/test/os-tests/tests/pf_key/acquire-compare.sh:44]: I'm not sure how to easily achieve that, we can't use `ping` since they don't have to _really_ be reachable, based on what you said.  Is there an easy way I'm not network-savvy enough to know?  I'm willing to leave it in the state it's in, for right now, since it's better than before.

Comment (danmcd) [reply usr/src/test/os-tests/tests/pf_key/acquire-compare.sh:44]: route(8) is your friend here but it's... potentially annoying in corner cases.  THere are other failures that might cut off the protect-with-IPsec path too.  And also, as I think about it, there might be lack of "-n" (no name resolution) in some commands.  I'll have to look at it more deeply later when I have cycles.

**Review by danmcd — APPROVED** [review]

**Comment by richlowe** [comment]
I merged this into the arm gate, but it still needs love upstream (or waiting for me)
